qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/RecallTestsIT.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.knn.bwc;
+
+import org.opensearch.knn.TestUtils;
+import org.opensearch.knn.index.SpaceType;
+
+import java.util.List;
+import java.util.Set;
+
+import static org.opensearch.knn.TestUtils.KNN_BWC_PREFIX;
+import static org.opensearch.knn.TestUtils.NODES_BWC_CLUSTER;
+import static org.opensearch.knn.index.KNNSettings.KNN_ALGO_PARAM_INDEX_THREAD_QTY;
+import static org.opensearch.knn.index.KNNSettings.KNN_MEMORY_CIRCUIT_BREAKER_ENABLED;
+
+public class RecallTestsIT extends AbstractRestartUpgradeTestCase {
+    private static final String TEST_INDEX_RECALL_OLD = KNN_BWC_PREFIX + "test-index-recall-old";
+    private static final String TEST_FIELD = "test-field";
+    private static final int DIMENSIONS = 50;
+    private static final int DIMENSIONS_RECALL_OLD = 1;
+    private static final int K = 10;
+    private static final int ADD_DOCS_CNT = 1000;
+    private static final int QUERY_COUNT = 100;
+
+    public void testKnnRecall() throws Exception {
+        waitForClusterHealthGreen(NODES_BWC_CLUSTER);
+        if (isRunningAgainstOldCluster()) {
+            createKnnIndex(testIndex, getKNNDefaultIndexSettings(), createKnnIndexMapping(TEST_FIELD, DIMENSIONS));
+            addDocs(testIndex, TEST_FIELD, DIMENSIONS, ADD_DOCS_CNT, true);
+
+            double recallVal = getkNNBWCRecallValue(testIndex, TEST_FIELD, ADD_DOCS_CNT, DIMENSIONS, QUERY_COUNT, K, true, SpaceType.L2);
+            createKnnIndex(TEST_INDEX_RECALL_OLD, getKNNDefaultIndexSettings(), createKnnIndexMapping(TEST_FIELD, DIMENSIONS_RECALL_OLD));
+            addKnnDoc(TEST_INDEX_RECALL_OLD, "1", TEST_FIELD, new Float[] { (float) recallVal });
+
+        } else {
+            double recallValUpgraded = getkNNBWCRecallValue(
+                testIndex,
+                TEST_FIELD,
+                ADD_DOCS_CNT,
+                DIMENSIONS,
+                QUERY_COUNT,
+                K,
+                true,
+                SpaceType.L2
+            );
```

Comment:
we definitely need to adopt some approach for easier usage of builder pattern for constructs like this one. of course it's not in scope for this change